### PR TITLE
Invalidate package objects transitively in name hashing

### LINKED ISSUE #2326: Name hashing breaks when using a package object hierarchy
- sbt version 0.13.9
- scala version 2.11.7

``` scala
// src/main/scala/p/impl/Foo.scala
package p.impl

trait Foo
```

``` scala
// src/main/scala/p/impl/PackageObject.scala
package p.impl

abstract class PackageObject extends Foo
```

``` scala
// src/main/scala/p/package.scala
import p.impl.PackageObject

package object p extends PackageObject
```

Any change in `Foo.scala`, even adding a space at the beginning of the file, causes:

```
[error] missing or invalid dependency detected while loading class file 'PackageObject.class'.
[error] Could not access type Foo in package p.impl,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'PackageObject.class' was compiled against an incompatible version of p.impl.
[error] one error found
[error] (compile:compileIncremental) Compilation failed
```

The current workaround is disabling name hashing:

```
incOptions ~= (_ withNameHashing false)
```

/cc @gkossakowski

[EDIT 2015-12-21: Moved `Foo` and `PackageObject` into a distinct package, and moved files into directories to follow the package structure.. to further remove doubts]

### PR DESCRIPTION
**Pull request by gkossakowski**
Modify `invalidatedPackageObjects` to look for package objects that
transitively inherit from an invalidated class instead of just inheriting
directly. This is the correct behavior that should have been implemented
in the first place. The bug comes from a subtle copy&paste mistake. The old
implementation used `publicInherited` relation and new one looks identical
but uses `inheritance` relation. The difference is that `publicInherited`
was a relation that was expanded transitively along inheritance chain but
`inheritance` is not. We have to perform the transitive walk in
`IncrementalNameHashing.invalidatedPackageObjects` implementation.

Mark `pkg-self` test as passing.

Fixes #2326

### CODE REVIEW

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

File: compile/inc/src/main/scala/sbt/inc/IncrementalNameHashing.scala [hunk 1]
```diff
@@ -18,7 +18,9 @@ private final class IncrementalNameHashing(log: Logger, options: IncOptions) ext
   // Package objects are fragile: if they inherit from an invalidated source, get "class file needed by package is missing" error
   //  This might be too conservative: we probably only need package objects for packages of invalidated sources.
   override protected def invalidatedPackageObjects(invalidated: Set[File], relations: Relations): Set[File] =
-    invalidated flatMap relations.inheritance.internal.reverse filter { _.getName == "package.scala" }
+    transitiveDeps(invalidated)(relations.inheritance.internal.reverse).filter {
+      _.getName == "package.scala"
+    }
```
Comment (paulp) [inline compile/inc/src/main/scala/sbt/inc/IncrementalNameHashing.scala:23]: I realize it's not new to this commit, but what is the basis for this filter? It's not supposed to be a correctness requirement that package objects be defined in files called package.scala. It is (according to the scala specification) a convention.

**Comment by Duhemm** [comment]
LGTM.

@paulp's comment is right. Maybe that we should add a flag `isPackageObject`, similar to `hasMacro`, and rely on that rather than the filename?

File: compile/inc/src/main/scala/sbt/inc/IncrementalNameHashing.scala (hunk 1 above)
Comment (gkossakowski) [inline compile/inc/src/main/scala/sbt/inc/IncrementalNameHashing.scala:23]: Yes, it was a quick hack. @Duhemm's idea would fix it for good.

**Comment by gkossakowski** [comment]
@Duhemm: yes, that's a good way to implement package object handling. Would you like to give it a try?

**Comment by Duhemm** [comment]
@gkossakowski Sure!

**Comment by gkossakowski** [comment]
great, small suggestion for implementation: introduce `hasPackageObject` to `APIs` that checks the flag in the `Source` object. This way the particular mechanism of storing that information can be hidden and changed later on. E.g. I'll need to change it for #1104.

**Comment by paulp** [comment]
Note if you haven't already that `_.getName == "package.scala"` occurs more than one place in the sbt source.

**Pull request merged**